test/e2e/upgrade_kubernetes_test.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/onsi/gomega/format"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Kubernetes (version) k8s_upgrade in target nodes [k8s_upgrade]", func() {
+	upgradeKubernetes()
+})
+
+func upgradeKubernetes() {
+	BeforeEach(func() {
+		osType := strings.ToLower(os.Getenv("OS"))
+		Expect(osType).ToNot(Equal(""))
+		validateGlobals(specName)
+
+		// We need to override clusterctl apply log folder to avoid getting our credentials exposed.
+		clusterctlLogFolder = filepath.Join(os.TempDir(), "clusters", bootstrapClusterProxy.GetName())
+	})
+	It("Should create a cluster and and run k8s_upgrade tests", func() {
+		By("Creating target cluster")
+		targetCluster = createTargetCluster(e2eConfig.GetVariable("FROM_K8S_VERSION"))
+
+		Logf("Starting Kubernetes upgrade tests")
+		var (
+			clusterClient       = bootstrapClusterProxy.GetClient()
+			targetClusterClient = targetCluster.GetClient()
+			clientSet           = targetCluster.GetClientSet()
+			kubernetesVersion   = e2eConfig.GetVariable("FROM_K8S_VERSION")
+			upgradedK8sVersion  = e2eConfig.GetVariable("KUBERNETES_VERSION")
+			controlplaneTaints  = []corev1.Taint{{Key: "node-role.kubernetes.io/control-plane", Effect: corev1.TaintEffectNoSchedule},
+				{Key: "node-role.kubernetes.io/master", Effect: corev1.TaintEffectNoSchedule}}
+			imageNamePrefix string
+		)
+
+		const (
+			artifactoryURL = "https://artifactory.nordix.org/artifactory/metal3/images/k8s"
+			imagesURL      = "http://172.22.0.1/images"
+			ironicImageDir = "/opt/metal3-dev-env/ironic/html/images"
+		)
+
+		Logf("KUBERNETES VERSION: %v", kubernetesVersion)
+		Logf("UPGRADED K8S VERSION: %v", upgradedK8sVersion)
+		Logf("NUMBER OF CONTROLPLANE BMH: %v", numberOfControlplane)
+		Logf("NUMBER OF WORKER BMH: %v", numberOfWorkers)
+
+		listBareMetalHosts(ctx, clusterClient, client.InNamespace(namespace))
+		listMetal3Machines(ctx, clusterClient, client.InNamespace(namespace))
+		listMachines(ctx, clusterClient, client.InNamespace(namespace))
+		listNodes(ctx, targetClusterClient)
+
+		// Download node image
+		By("Download image")
+		osType := strings.ToLower(os.Getenv("OS"))
+		Expect(osType).ToNot(Equal(""))
+		imageNamePrefix = "UBUNTU_22.04_NODE_IMAGE_K8S"
+		imageName := fmt.Sprintf("%s_%s.qcow2", imageNamePrefix, upgradedK8sVersion)
+		Logf("IMAGE_NAME: %v", imageName)
+		rawImageName := fmt.Sprintf("%s_%s-raw.img", imageNamePrefix, upgradedK8sVersion)
+		Logf("RAW_IMAGE_NAME: %v", rawImageName)
+		imageLocation := fmt.Sprintf("%s_%s/", artifactoryURL, upgradedK8sVersion)
+		Logf("IMAGE_LOCATION: %v", imageLocation)
+		imageURL := fmt.Sprintf("%s/%s", imagesURL, rawImageName)
+		Logf("IMAGE_URL: %v", imageURL)
+		imageChecksum := fmt.Sprintf("%s/%s.md5sum", imagesURL, rawImageName)
+		Logf("IMAGE_CHECKSUM: %v", imageChecksum)
```

Comment:
can we use this function https://github.com/metal3-io/cluster-api-provider-metal3/blob/496bfb481f8754772e8fe77fbda4464f0c339255/test/e2e/common.go#L93

Reply:
done